Buildsystem improvements for the Mono module

neikeq opened the pull request:
- Make sure to search the mono installation directory for the right architecture in the windows registry.
- Do not build GodotSharpTools directly to #bin dir. Instead build to the default output path and copy it. This way we avoid MSBuild adding files we don't want to #bin.
- Add hint path for MSBuild in OSX.
- Copy shared library on Unix if not statically linking.
- Use vswhere to search MSBuild and search for 14.0 tools version in the registry instead of 4.0.
- SCons will only fallback xbuild when msbuild is not found if 'xbuild_fallback=yes' is passed to the command.
- Use mono's assembly path as FrameworkPathOverride if using with system's MSBuild (not mono's fork).
- Cleanup.

neikeq committed:
Buildsystem improvements for the Mono module
- Make sure to search the mono installation directory for the right architecture in the windows registry.
- Do not build GodotSharpTools directly to #bin dir. Instead build to the default output path and copy it. This way we avoid MSBuild adding files we don't want to #bin.
- Add hint path for MSBuild in OSX.
- Copy shared library on Unix if not statically linking.
- Use vswhere to search MSBuild and search for 14.0 tools version in the registry instead of 4.0.
- SCons will only fallback xbuild when msbuild is not found if 'xbuild_fallback=yes' is passed to the command.
- Use mono's assembly path as FrameworkPathOverride if using with system's MSBuild (not mono's fork).
- Cleanup.

neikeq commented on modules/mono/SCsub:
```diff
@@ -1,121 +1,150 @@
-#!/usr/bin/env python
```
Not sure what's wrong with this diff...

NathanWarden replied:
Line endings?

akien-mga replied:
Yes, that's the file before and after your patch:
```
$ file modules/mono/SCsub 
modules/mono/SCsub: Python script, ASCII text executable
$ git apply ~/Downloads/df950b9a2950d854d0c14de966856d6296fa8dcb.patch
[...]
$ file modules/mono/SCsub 
modules/mono/SCsub: Python script, ASCII text executable, with CRLF line terminators
```

akien-mga commented:
This change will make it difficult for me to make alpha2 binaries, since I can't find a way to build MSBuild from source and I need to be able to generate the glue.

akien-mga commented:
BTW, regardless of whether we keep support for xbuild for now, `modules/mono/SCsub` should be made to check for msbuild/xbuild in `$MONO64_PREFIX/bin` and `$MONO32_PREFIX/bin` too (and that by default, so that they can override system installed older versions).

neikeq commented:
Updated. In order to build with xbuild if msbuild is not found, you will have to pass `xbuild_fallback=yes` to scons. For the editor, I changed the setting name to `xbuild (Deprecated)` and it's not displayed on Windows.
